### ISSUE TITLE
KJ Test failures - MSVC C++17

### ISSUE THREAD

**Issue by uniflare** [capnproto/capnproto#1902]
With this in my toolchain for master branch:

```
set(CMAKE_SYSTEM_VERSION 10.0)

set(CMAKE_C_STANDARD 11)
set(CMAKE_C_STANDARD_REQUIRED ON)
set(CMAKE_CXX_STANDARD 17)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
set(CMAKE_CXX_EXTENSIONS OFF)

set(CMAKE_CXX_FLAGS             "/nologo /MP /W4 /GF /Gy /utf-8 /bigobj /EHsc"   CACHE STRING "C++ Global Flags" FORCE)
set(CMAKE_COMPILE_FLAGS_DEBUG   "/MDd /GS /RTC1 /DDEBUG /Zi /Zo /Od /Ob0"        CACHE STRING "C++ Debug Flags" FORCE)
set(CMAKE_COMPILE_FLAGS_RELEASE "/MD /GS- /DNDEBUG /Zo- /Ox"                     CACHE STRING "C++ Release Flags" FORCE)
set(CMAKE_LINK_FLAGS_DEBUG      "/DEBUG /OPT:NOREF /OPT:NOICF"                   CACHE STRING "Linker Debug Flags" FORCE)
set(CMAKE_LINK_FLAGS_RELEASE    "/DEBUG:NONE /OPT:REF /OPT:ICF /INCREMENTAL:NO"  CACHE STRING "Linker Release Flags" FORCE)
```

I get these failures when running kj-tests:

```
kj\common-test.c++:635: failed: expected !(f == f)
stack: 7ff6eed01c44 7ff6eeb1e055 7ff6eeb3d128 7ff6eecdc742 7ff6eecdac69 7ff6eecdf38b 7ff6eecd818a 7ff6eecd7f7b 7ff6eecdc853 7ff6eed30e66 7ff6eed3440b 7ff6eed31155 7ff6eed30f98 7ff6eed30c39 7ff6eed2a539 7ff6eed3cdce 7ff6eece091d 7ff6eed4e968 7ff6eed4e84d 7ff6eed4e70d 7ff6eed4e9fd 7ffc2a94257c 7ffc2c54aa57
[ FAIL ] common-test.c++:612: legacy test: Common/MinMaxValue (2407 ╬╝s)

kj\common-test.c++:817: failed: expected ArrayPtr<double>(d, 1) != ArrayPtr<double>(d, 1) [inf != inf]
stack: 7ff6eed01c44 7ff6eeb25205 7ff6eeb4e668 7ff6eecdc742 7ff6eecdac69 7ff6eecdf38b 7ff6eecd818a 7ff6eecd7f7b 7ff6eecdc853 7ff6eed30e66 7ff6eed3440b 7ff6eed31155 7ff6eed30f98 7ff6eed30c39 7ff6eed2a539 7ff6eed3cdce 7ff6eece091d 7ff6eed4e968 7ff6eed4e84d 7ff6eed4e70d 7ff6eed4e9fd 7ffc2a94257c 7ffc2c54aa57
[ FAIL ] common-test.c++:795: ArrayPtr operator == (588 ╬╝s)

kj\string-test.c++:100: failed: expected isNaN(StringPtr("nan").parseAs<double>())
stack: 7ff6eed01c44 7ff6eeb247d5 7ff6eebce131 7ff6eecdc742 7ff6eecdac69 7ff6eecdf38b 7ff6eecd818a 7ff6eecd7f7b 7ff6eecdc853 7ff6eed30e66 7ff6eed3440b 7ff6eed31155 7ff6eed30f98 7ff6eed30c39 7ff6eed2a539 7ff6eed3cdce 7ff6eece091d 7ff6eed4e968 7ff6eed4e84d 7ff6eed4e70d 7ff6eed4e9fd 7ffc2a94257c 7ffc2c54aa57
kj\string-test.c++:101: failed: expected isNaN(StringPtr("NAN").parseAs<double>())
stack: 7ff6eed01c44 7ff6eeb247d5 7ff6eebce1eb 7ff6eecdc742 7ff6eecdac69 7ff6eecdf38b 7ff6eecd818a 7ff6eecd7f7b 7ff6eecdc853 7ff6eed30e66 7ff6eed3440b 7ff6eed31155 7ff6eed30f98 7ff6eed30c39 7ff6eed2a539 7ff6eed3cdce 7ff6eece091d 7ff6eed4e968 7ff6eed4e84d 7ff6eed4e70d 7ff6eed4e9fd 7ffc2a94257c 7ffc2c54aa57
kj\string-test.c++:102: failed: expected isNaN(StringPtr("NaN").parseAs<double>())
stack: 7ff6eed01c44 7ff6eeb247d5 7ff6eebce2a5 7ff6eecdc742 7ff6eecdac69 7ff6eecdf38b 7ff6eecd818a 7ff6eecd7f7b 7ff6eecdc853 7ff6eed30e66 7ff6eed3440b 7ff6eed31155 7ff6eed30f98 7ff6eed30c39 7ff6eed2a539 7ff6eed3cdce 7ff6eece091d 7ff6eed4e968 7ff6eed4e84d 7ff6eed4e70d 7ff6eed4e9fd 7ffc2a94257c 7ffc2c54aa57
[ FAIL ] string-test.c++:84: legacy test: String/parseAs (16312 ╬╝s)

kj\string-test.c++:185: failed: expected isNaN(StringPtr("nan").tryParseAs<double>().orDefault(0.0)) == true [false == true]
stack: 7ff6eed01c44 7ff6eebabbb5 7ff6eebbf148 7ff6eecdc742 7ff6eecdac69 7ff6eecdf38b 7ff6eecd818a 7ff6eecd7f7b 7ff6eecdc853 7ff6eed30e66 7ff6eed3440b 7ff6eed31155 7ff6eed30f98 7ff6eed30c39 7ff6eed2a539 7ff6eed3cdce 7ff6eece091d 7ff6eed4e968 7ff6eed4e84d 7ff6eed4e70d 7ff6eed4e9fd 7ffc2a94257c 7ffc2c54aa57
kj\string-test.c++:186: failed: expected isNaN(StringPtr("NAN").tryParseAs<double>().orDefault(0.0)) == true [false == true]
stack: 7ff6eed01c44 7ff6eebabbb5 7ff6eebbf2a6 7ff6eecdc742 7ff6eecdac69 7ff6eecdf38b 7ff6eecd818a 7ff6eecd7f7b 7ff6eecdc853 7ff6eed30e66 7ff6eed3440b 7ff6eed31155 7ff6eed30f98 7ff6eed30c39 7ff6eed2a539 7ff6eed3cdce 7ff6eece091d 7ff6eed4e968 7ff6eed4e84d 7ff6eed4e70d 7ff6eed4e9fd 7ffc2a94257c 7ffc2c54aa57
kj\string-test.c++:187: failed: expected isNaN(StringPtr("NaN").tryParseAs<double>().orDefault(0.0)) == true [false == true]
stack: 7ff6eed01c44 7ff6eebabbb5 7ff6eebbf404 7ff6eecdc742 7ff6eecdac69 7ff6eecdf38b 7ff6eecd818a 7ff6eecd7f7b 7ff6eecdc853 7ff6eed30e66 7ff6eed3440b 7ff6eed31155 7ff6eed30f98 7ff6eed30c39 7ff6eed2a539 7ff6eed3cdce 7ff6eece091d 7ff6eed4e968 7ff6eed4e84d 7ff6eed4e70d 7ff6eed4e9fd 7ffc2a94257c 7ffc2c54aa57
[ FAIL ] string-test.c++:168: legacy test: String/tryParseAs (1747 ╬╝s)

kj\string-test.c++:312: failed: expected memcmp(&parsedNan, &canonicalNan, sizeof(parsedNan)) == 0 [1 == 0]
stack: 7ff6eed01c44 7ff6eebaa545 7ff6eebc7e99 7ff6eecdc742 7ff6eecdac69 7ff6eecdf38b 7ff6eecd818a 7ff6eecd7f7b 7ff6eecdc853 7ff6eed30e66 7ff6eed3440b 7ff6eed31155 7ff6eed30f98 7ff6eed30c39 7ff6eed2a539 7ff6eed3cdce 7ff6eece091d 7ff6eed4e968 7ff6eed4e84d 7ff6eed4e70d 7ff6eed4e9fd 7ffc2a94257c 7ffc2c54aa57
kj\string-test.c++:317: failed: expected memcmp(&parsedNan, &canonicalNan, sizeof(parsedNan)) == 0 [1 == 0]
stack: 7ff6eed01c44 7ff6eebaa545 7ff6eebc7fb0 7ff6eecdc742 7ff6eecdac69 7ff6eecdf38b 7ff6eecd818a 7ff6eecd7f7b 7ff6eecdc853 7ff6eed30e66 7ff6eed3440b 7ff6eed31155 7ff6eed30f98 7ff6eed30c39 7ff6eed2a539 7ff6eed3cdce 7ff6eece091d 7ff6eed4e968 7ff6eed4e84d 7ff6eed4e70d 7ff6eed4e9fd 7ffc2a94257c 7ffc2c54aa57
[ FAIL ] string-test.c++:306: parsing 'nan' returns canonical NaN value (1078 ╬╝s)
```

**Comment by kentonv** [comment]
Per your comment in the other thread, it sounds like these failures are a result of `/fp:fast`. This flag enables optimizations that may assume special values like infinity and NaN don't ever appear, breaking the cases where they do. As one example, it appears the compiler has decided that `f == f` is always true -- since the only case where it isn't true is, of course, NaN.

I think you should be OK as long as you don't ever try to use infinity or NaN in your program. However, if you accept Cap'n Proto messages from untrusted sources which contain floating point values, it's hard to say what will happen if someone sends you infinitiy or NaN. Probably you just get garbage, I suppose? But in the worst case it might be possible to trigger undefined behavior... that _seems_ unlikely but hard for me to rule out.

In short, this flag which is explicitly documented as breaking things, broke things. This is not a bug in KJ, it's a risk you take when enabling this flag. It's up to you to convince yourself that it won't be an issue.

In any case I don't think we should change anything about the tests since they are legitimately broken in this mode.

**Comment by uniflare** [comment]
Thanks for the explanation :)
Indeed it was a dirty workspace where fp:fast was still enabled.
